

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bkbit.data_translators.genome_annotation_translator &mdash; bkbit  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            bkbit
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">GETTING STARTED</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">DATA TRANSLATORS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../specimen_file_manifest.html">Specimen Digital File Manifest</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../list_specimen_library_aliquot.html">Specimen Library Aliquot List</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../specimen_metadata.html">Specimen Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../genome_annotation.html">Annotated Genome Data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">MODEL CONVERTERS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../spreadsheet_converter.html">Spreadsheet to LinkML Schema</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">MODEL EDITORS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../linkml_trimmer.html">LinkML Schema Trimmer</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">REFERENCE</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">bkbit</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">bkbit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bkbit.data_translators.genome_annotation_translator</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bkbit.data_translators.genome_annotation_translator</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module for downloading, parsing, and processing GFF3 files from NCBI and Ensembl repositories. This module provides functionality to:</span>

<span class="sd">1. Download a GFF3 file from a specified URL and calculate its checksums.</span>
<span class="sd">2. Parse the GFF3 file to extract gene annotations.</span>
<span class="sd">3. Generate various metadata objects such as organism taxon, genome assembly, and genome annotation.</span>
<span class="sd">4. Serialize the extracted information into JSON-LD format for further use.</span>

<span class="sd">Classes:</span>
<span class="sd">    Gff3: The Gff3 class is designed to handle the complete lifecycle of downloading, parsing, and processing GFF3 files from NCBI or Ensembl repositories. It extracts gene annotations and serializes the data into JSON-LD format.</span>

<span class="sd">Functions:</span>
<span class="sd">    gff2jsonld: The gff2jsonld function is responsible for creating GeneAnnotation objects from a provided GFF3 file and serializing the extracted information into the JSON-LD format.</span>

<span class="sd">Usage:</span>
<span class="sd">    The module can be run as a standalone script by executing it with appropriate arguments and options:</span>
<span class="sd">    </span>
<span class="sd">    ```</span>
<span class="sd">    python genome_annotation_translator.py &lt;content_url&gt; -a &lt;assembly_accession&gt; -s &lt;assembly_strain&gt; -l &lt;log_level&gt; -f</span>
<span class="sd">    ```</span>
<span class="sd">    </span>
<span class="sd">    The script will download the GFF3 file from the specified URL, parse it, and serialize the extracted information into JSON-LD format.</span>

<span class="sd">Example:</span>
<span class="sd">    ```</span>
<span class="sd">    python genome_annotation_translator.py &quot;https://example.com/path/to/gff3.gz&quot; -a &quot;GCF_000001405.39&quot; -s &quot;strain_name&quot; -l &quot;INFO&quot; -f True</span>
<span class="sd">    ```</span>
<span class="sd">    </span>
<span class="sd">Dependencies:</span>
<span class="sd">    - re</span>
<span class="sd">    - hashlib</span>
<span class="sd">    - tempfile</span>
<span class="sd">    - uuid</span>
<span class="sd">    - urllib</span>
<span class="sd">    - urllib.request</span>
<span class="sd">    - urllib.parse</span>
<span class="sd">    - os</span>
<span class="sd">    - json</span>
<span class="sd">    - datetime</span>
<span class="sd">    - collections.defaultdict</span>
<span class="sd">    - subprocess</span>
<span class="sd">    - gzip</span>
<span class="sd">    - tqdm</span>
<span class="sd">    - click</span>
<span class="sd">    - pkg_resources</span>
<span class="sd">    - bkbit.models.genome_annotation as ga</span>
<span class="sd">    - bkbit.utils.setup_logger as setup_logger</span>
<span class="sd">    - bkbit.utils.load_json as load_json    </span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">hashlib</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">uuid</span>
<span class="kn">import</span> <span class="nn">urllib</span>
<span class="kn">import</span> <span class="nn">urllib.request</span>
<span class="kn">from</span> <span class="nn">urllib.parse</span> <span class="kn">import</span> <span class="n">urlparse</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">click</span>
<span class="kn">import</span> <span class="nn">pkg_resources</span>
<span class="kn">from</span> <span class="nn">bkbit.models</span> <span class="kn">import</span> <span class="n">genome_annotation</span> <span class="k">as</span> <span class="n">ga</span>
<span class="kn">from</span> <span class="nn">bkbit.utils.setup_logger</span> <span class="kn">import</span> <span class="n">setup_logger</span>
<span class="kn">from</span> <span class="nn">bkbit.utils.load_json</span> <span class="kn">import</span> <span class="n">load_json</span>



<span class="c1">## CONSTANTS ##</span>

<span class="n">PREFIX_MAP</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;NCBITaxon&quot;</span><span class="p">:</span> <span class="s2">&quot;http://purl.obolibrary.org/obo/NCBITaxon_&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NCBIGene&quot;</span><span class="p">:</span> <span class="s2">&quot;http://identifiers.org/ncbigene/&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ENSEMBL&quot;</span><span class="p">:</span> <span class="s2">&quot;http://identifiers.org/ensembl/&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NCBIAssembly&quot;</span><span class="p">:</span> <span class="s2">&quot;https://www.ncbi.nlm.nih.gov/assembly/&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">NCBI_GENE_ID_PREFIX</span> <span class="o">=</span> <span class="s2">&quot;NCBIGene&quot;</span>
<span class="n">ENSEMBL_GENE_ID_PREFIX</span> <span class="o">=</span> <span class="s2">&quot;ENSEMBL&quot;</span>
<span class="n">TAXON_PREFIX</span> <span class="o">=</span> <span class="s2">&quot;NCBITaxon&quot;</span>
<span class="n">ASSEMBLY_PREFIX</span> <span class="o">=</span> <span class="s2">&quot;NCBIAssembly&quot;</span>
<span class="n">BICAN_ANNOTATION_PREFIX</span> <span class="o">=</span> <span class="s2">&quot;bican:annotation-&quot;</span>
<span class="n">GENOME_ANNOTATION_DESCRIPTION_FORMAT</span> <span class="o">=</span> <span class="p">(</span>
    <span class="s2">&quot;</span><span class="si">{authority}</span><span class="s2"> </span><span class="si">{taxon_scientific_name}</span><span class="s2"> Annotation Release </span><span class="si">{genome_version}</span><span class="s2">&quot;</span>
<span class="p">)</span>
<span class="n">DEFAULT_FEATURE_FILTER</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="s2">&quot;ncRNA_gene&quot;</span><span class="p">)</span>
<span class="n">DEFAULT_HASH</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;MD5&quot;</span><span class="p">,)</span>
<span class="n">LOG_FILE_NAME</span> <span class="o">=</span> <span class="p">(</span>
    <span class="s2">&quot;gff3_translator_&quot;</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2">_%H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.log&quot;</span>
<span class="p">)</span>
<span class="n">TAXON_DIR_PATH</span> <span class="o">=</span> <span class="s2">&quot;../utils/ncbi_taxonomy/&quot;</span>
<span class="n">SCIENTIFIC_NAME_TO_TAXONID_PATH</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="n">TAXON_DIR_PATH</span> <span class="o">+</span> <span class="s2">&quot;scientific_name_to_taxid.json&quot;</span><span class="p">)</span>
<span class="n">TAXON_SCIENTIFIC_NAME_PATH</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="n">TAXON_DIR_PATH</span> <span class="o">+</span> <span class="s2">&quot;taxid_to_scientific_name.json&quot;</span><span class="p">)</span>
<span class="n">TAXON_COMMON_NAME_PATH</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="n">TAXON_DIR_PATH</span> <span class="o">+</span> <span class="s2">&quot;taxid_to_common_name.json&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="Gff3">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3">[docs]</a>
<span class="k">class</span> <span class="nc">Gff3</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The Gff3 class is responsible for downloading, parsing, and processing of GFF3 files from NCBI and Ensembl repositories.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        content_url (str): The URL of the GFF file.</span>
<span class="sd">        assembly_accession (str): The ID of the genome assembly.</span>
<span class="sd">        assembly_strain (str, optional): The strain of the genome assembly. Defaults to None.</span>
<span class="sd">        log_level (str): The logging level. Defaults to &#39;WARNING&#39;.</span>
<span class="sd">        log_to_file (bool): Flag to log messages to a file. Defaults to False.</span>

<span class="sd">    Methods:</span>
<span class="sd">        __init__(content_url, assembly_accession=None, assembly_strain=None, log_level=&quot;WARNING&quot;, log_to_file=False):</span>
<span class="sd">            Initializes the Gff3 class with the provided parameters.</span>

<span class="sd">        parse_url():</span>
<span class="sd">            Parses the content URL and extracts information about the genome annotation.</span>

<span class="sd">        __download_gff_file():</span>
<span class="sd">            Downloads a GFF file from a given URL and calculates the MD5, SHA256, and SHA1 hashes.</span>

<span class="sd">        generate_organism_taxon(taxon_id):</span>
<span class="sd">            Generates an organism taxon object based on the provided taxon ID.</span>

<span class="sd">        assign_authority_type(authority):</span>
<span class="sd">            Assigns the authority type based on the given authority string.</span>

<span class="sd">        generate_genome_assembly(assembly_id, assembly_version, assembly_label, assembly_strain=None):</span>
<span class="sd">            Generates a genome assembly object based on the provided parameters.</span>

<span class="sd">        generate_genome_annotation(genome_label, genome_version):</span>
<span class="sd">            Generates a genome annotation object based on the provided parameters.</span>

<span class="sd">        generate_digest(hash_values, hash_functions=DEFAULT_HASH):</span>
<span class="sd">            Generates checksum digests for the GFF file using the specified hash functions.</span>

<span class="sd">        __get_line_count(file_path):</span>
<span class="sd">            Returns the line count of a file.</span>

<span class="sd">        parse(feature_filter=DEFAULT_FEATURE_FILTER):</span>
<span class="sd">            Parses the GFF file and extracts gene annotations based on the provided feature filter.</span>

<span class="sd">        generate_ensembl_gene_annotation(attributes, curr_line_num):</span>
<span class="sd">            Generates a GeneAnnotation object for Ensembl based on the provided attributes.</span>

<span class="sd">        generate_ncbi_gene_annotation(attributes, curr_line_num):</span>
<span class="sd">            Generates a GeneAnnotation object for NCBI based on the provided attributes.</span>

<span class="sd">        __get_attribute(attributes, attribute_name, curr_line_num):</span>
<span class="sd">            Retrieves the value of a specific attribute from the given attributes dictionary.</span>

<span class="sd">        __resolve_ncbi_gene_annotation(new_gene_annotation, curr_line_num):</span>
<span class="sd">            Resolves conflicts between existing and new gene annotations based on certain conditions.</span>

<span class="sd">        __merge_values(t):</span>
<span class="sd">            Merges values from a list of lists into a dictionary of sets.</span>

<span class="sd">        serialize_to_jsonld(exclude_none=True, exclude_unset=False):</span>
<span class="sd">            Serializes the object and either writes it to the specified output file or prints it to the CLI.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="Gff3.__init__">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">content_url</span><span class="p">,</span>
        <span class="n">assembly_accession</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">assembly_strain</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
        <span class="n">log_to_file</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes an instance of the GFFTranslator class.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        - content_url (str): The URL of the GFF file.</span>
<span class="sd">        - assembly_id (str): The ID of the genome assembly.</span>
<span class="sd">        - assembly_strain (str, optional): The strain of the genome assembly. Defaults to None.</span>
<span class="sd">        - hash_functions (tuple[str]): A tuple of hash functions to use for generating checksums. Defaults to (&#39;MD5&#39;).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logger</span><span class="p">(</span><span class="n">LOG_FILE_NAME</span><span class="p">,</span> <span class="n">log_level</span><span class="p">,</span> <span class="n">log_to_file</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">scientific_name_to_taxonid</span> <span class="o">=</span> <span class="n">load_json</span><span class="p">(</span><span class="n">SCIENTIFIC_NAME_TO_TAXONID_PATH</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxon_scientific_name</span> <span class="o">=</span> <span class="n">load_json</span><span class="p">(</span><span class="n">TAXON_SCIENTIFIC_NAME_PATH</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxon_common_name</span> <span class="o">=</span> <span class="n">load_json</span><span class="p">(</span><span class="n">TAXON_COMMON_NAME_PATH</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s2">&quot;NCBI Taxonomy not downloaded. Run &#39;bkbit download-ncbi-taxonomy&#39; command first.&quot;</span> <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">content_url</span> <span class="o">=</span> <span class="n">content_url</span>

        <span class="c1">## STEP 1: Parse the content URL to get metadata</span>
        <span class="c1"># Parse content_url to get metadata</span>
        <span class="n">url_metadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parse_url</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">url_metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span>
                <span class="s2">&quot;The provided content URL is not supported. Please provide a valid URL.&quot;</span>
            <span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;The provided content URL is not supported. Please provide a valid URL.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Define variables to store metadata</span>
        <span class="p">(</span>
            <span class="n">taxon_id</span><span class="p">,</span>
            <span class="n">assembly_id</span><span class="p">,</span>
            <span class="n">assembly_version</span><span class="p">,</span>
            <span class="n">assembly_label</span><span class="p">,</span>
            <span class="n">genome_label</span><span class="p">,</span>
            <span class="n">genome_version</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">=</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="c1"># Assign the authority type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">authority</span> <span class="o">=</span> <span class="n">url_metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;authority&quot;</span><span class="p">)</span>

        <span class="c1"># Assign the taxon_id and assembly_id based on the authority</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">authority</span><span class="o">.</span><span class="n">value</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">NCBI</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">url_metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;taxonid&quot;</span><span class="p">)</span>
            <span class="n">assembly_id</span> <span class="o">=</span> <span class="n">url_metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_accession&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">authority</span><span class="o">.</span><span class="n">value</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">ENSEMBL</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="n">taxon_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scientific_name_to_taxonid</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                <span class="n">url_metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;scientific_name&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">assembly_accession</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span>
                    <span class="s2">&quot;The assembly ID is required for Ensembl URLs. Please provide the assembly ID.&quot;</span>
                <span class="p">)</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;The assembly ID is required for Ensembl URLs. Please provide the assembly ID.&quot;</span>
                <span class="p">)</span>
            <span class="n">assembly_id</span> <span class="o">=</span> <span class="n">assembly_accession</span>

        <span class="c1"># Assign assembly_version, assembly_label, genome_version, and genome_label</span>
        <span class="n">assembly_version</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">assembly_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">assembly_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="n">assembly_label</span> <span class="o">=</span> <span class="n">url_metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">)</span>
        <span class="n">genome_version</span> <span class="o">=</span> <span class="n">url_metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;release_version&quot;</span><span class="p">)</span>
        <span class="n">genome_label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">authority</span><span class="o">.</span><span class="n">value</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">+</span> <span class="n">taxon_id</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">+</span> <span class="n">genome_version</span>

        <span class="c1">## STEP 2: Download the GFF file</span>
        <span class="c1"># Download the GFF file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gff_file</span><span class="p">,</span> <span class="n">hash_values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__download_gff_file</span><span class="p">()</span>

        <span class="c1">## STEP 3: Generate the organism taxon, genome assembly, checksums, and genome annotation objects</span>
        <span class="c1"># Generate the organism taxon object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_organism_taxon</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_assembly</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_genome_assembly</span><span class="p">(</span>
            <span class="n">assembly_id</span><span class="p">,</span> <span class="n">assembly_version</span><span class="p">,</span> <span class="n">assembly_label</span><span class="p">,</span> <span class="n">assembly_strain</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">checksums</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_digest</span><span class="p">(</span><span class="n">hash_values</span><span class="p">,</span> <span class="n">DEFAULT_HASH</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_annotation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_genome_annotation</span><span class="p">(</span>
            <span class="n">genome_label</span><span class="p">,</span> <span class="n">genome_version</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span> <span class="o">=</span> <span class="p">{}</span></div>


<div class="viewcode-block" id="Gff3.parse_url">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.parse_url">[docs]</a>
    <span class="k">def</span> <span class="nf">parse_url</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parses the content URL and extracts information about the genome annotation.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A dictionary containing the following information:</span>
<span class="sd">            - &#39;authority&#39;: The authority type (NCBI or ENSEMBL).</span>
<span class="sd">            - &#39;taxonid&#39;: The taxon ID of the genome.</span>
<span class="sd">            - &#39;release_version&#39;: The release version of the genome annotation.</span>
<span class="sd">            - &#39;assembly_accession&#39;: The assembly accession of the genome.</span>
<span class="sd">            - &#39;assembly_name&#39;: The name of the assembly.</span>
<span class="sd">            - &#39;species&#39;: The species name (only for ENSEMBL URLs).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Define regex patterns for NCBI and Ensembl URLs</span>
        <span class="c1"># NCBI : [assembly accession.version]_[assembly name]_[content type].[optional format]</span>
        <span class="c1"># ENSEMBL :  &lt;species&gt;.&lt;assembly&gt;.&lt;_version&gt;.gff3.gz -&gt; organism full name, assembly name, genome version</span>
        <span class="n">ncbi_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;/genomes/all/annotation_releases/(\d+)(?:/(\d+))?/(GCF_\d+\.\d+)[_-]([^/]+)/(GCF_\d+\.\d+)[_-]([^/]+)_genomic\.gff\.gz&quot;</span>
        <span class="n">ensembl_pattern</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">r</span><span class="s2">&quot;/pub/release-(\d+)/gff3/([^/]+)/([^/.]+)\.([^/.]+)\.([^/.]+)\.gff3\.gz&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Parse the URL to get the path</span>
        <span class="n">parsed_url</span> <span class="o">=</span> <span class="n">urlparse</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">content_url</span><span class="p">)</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">parsed_url</span><span class="o">.</span><span class="n">path</span>

        <span class="c1"># Determine if the URL is from NCBI or Ensembl and extract information</span>
        <span class="k">if</span> <span class="s2">&quot;ncbi&quot;</span> <span class="ow">in</span> <span class="n">parsed_url</span><span class="o">.</span><span class="n">netloc</span><span class="p">:</span>
            <span class="n">ncbi_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">ncbi_pattern</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ncbi_match</span><span class="p">:</span>
                <span class="k">return</span> <span class="p">{</span>
                    <span class="s2">&quot;authority&quot;</span><span class="p">:</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">NCBI</span><span class="p">,</span>
                    <span class="s2">&quot;taxonid&quot;</span><span class="p">:</span> <span class="n">ncbi_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span>
                    <span class="s2">&quot;release_version&quot;</span><span class="p">:</span> <span class="p">(</span>
                        <span class="n">ncbi_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">ncbi_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                        <span class="k">else</span> <span class="n">ncbi_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
                    <span class="p">),</span>
                    <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="n">ncbi_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span>
                    <span class="s2">&quot;assembly_name&quot;</span><span class="p">:</span> <span class="n">ncbi_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">6</span><span class="p">),</span>
                <span class="p">}</span>

        <span class="k">elif</span> <span class="s2">&quot;ensembl&quot;</span> <span class="ow">in</span> <span class="n">parsed_url</span><span class="o">.</span><span class="n">netloc</span><span class="p">:</span>
            <span class="n">ensembl_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">ensembl_pattern</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ensembl_match</span><span class="p">:</span>
                <span class="k">return</span> <span class="p">{</span>
                    <span class="s2">&quot;authority&quot;</span><span class="p">:</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">ENSEMBL</span><span class="p">,</span>
                    <span class="s2">&quot;release_version&quot;</span><span class="p">:</span> <span class="n">ensembl_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span>
                    <span class="s2">&quot;scientific_name&quot;</span><span class="p">:</span> <span class="n">ensembl_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span>
                    <span class="s2">&quot;assembly_name&quot;</span><span class="p">:</span> <span class="n">ensembl_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">4</span><span class="p">),</span>
                <span class="p">}</span>

        <span class="c1"># If no match is found, return None</span>
        <span class="k">return</span> <span class="kc">None</span></div>


    <span class="k">def</span> <span class="nf">__download_gff_file</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Downloads a GFF file from a given URL and calculates the MD5, SHA256, and SHA1 hashes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: A tuple containing the path to the downloaded gzip file and a dictionary</span>
<span class="sd">            with the MD5, SHA256, and SHA1 hashes of the file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">content_url</span><span class="p">)</span>
        <span class="n">total_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">headers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;content-length&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
        <span class="n">block_size</span> <span class="o">=</span> <span class="mi">1024</span>  <span class="c1"># 1 Kilobyte</span>

        <span class="c1"># Create hash objects</span>
        <span class="n">md5_hash</span> <span class="o">=</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">()</span>
        <span class="n">sha256_hash</span> <span class="o">=</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">sha256</span><span class="p">()</span>
        <span class="n">sha1_hash</span> <span class="o">=</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">sha1</span><span class="p">()</span>

        <span class="c1"># Create a temporary file for the gzip data</span>
        <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.gz&quot;</span><span class="p">,</span> <span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_gzip</span><span class="p">:</span>
            <span class="n">gzip_file_path</span> <span class="o">=</span> <span class="n">f_gzip</span><span class="o">.</span><span class="n">name</span>

            <span class="c1"># Create a progress bar</span>
            <span class="n">progress_bar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span>
                <span class="n">total</span><span class="o">=</span><span class="n">total_size</span><span class="p">,</span>
                <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;iB&quot;</span><span class="p">,</span>
                <span class="n">unit_scale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Downloading GFF file&quot;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Read the file in chunks, write to the temporary file, and update the hash</span>
            <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">block_size</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">data</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="n">f_gzip</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                <span class="n">md5_hash</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                <span class="n">sha256_hash</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                <span class="n">sha1_hash</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                <span class="n">progress_bar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>

            <span class="n">progress_bar</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1"># Return the path to the temporary file and the md5 hash</span>
        <span class="k">return</span> <span class="n">gzip_file_path</span><span class="p">,</span> <span class="p">{</span>
            <span class="s2">&quot;MD5&quot;</span><span class="p">:</span> <span class="n">md5_hash</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">(),</span>
            <span class="s2">&quot;SHA256&quot;</span><span class="p">:</span> <span class="n">sha256_hash</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">(),</span>
            <span class="s2">&quot;SHA1&quot;</span><span class="p">:</span> <span class="n">sha1_hash</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">(),</span>
        <span class="p">}</span>

<div class="viewcode-block" id="Gff3.generate_organism_taxon">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.generate_organism_taxon">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_organism_taxon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates an organism taxon object based on the provided taxon ID.</span>

<span class="sd">        Args:</span>
<span class="sd">            taxon_id (str): The taxon ID of the organism.</span>

<span class="sd">        Returns:</span>
<span class="sd">            ga.OrganismTaxon: The generated organism taxon object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">ga</span><span class="o">.</span><span class="n">OrganismTaxon</span><span class="p">(</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">TAXON_PREFIX</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">taxon_id</span><span class="p">,</span>
            <span class="n">full_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">taxon_scientific_name</span><span class="p">[</span><span class="n">taxon_id</span><span class="p">],</span>
            <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">taxon_common_name</span><span class="p">[</span><span class="n">taxon_id</span><span class="p">],</span>
            <span class="n">iri</span><span class="o">=</span><span class="n">PREFIX_MAP</span><span class="p">[</span><span class="n">TAXON_PREFIX</span><span class="p">]</span> <span class="o">+</span> <span class="n">taxon_id</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Gff3.assign_authority_type">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.assign_authority_type">[docs]</a>
    <span class="k">def</span> <span class="nf">assign_authority_type</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">authority</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Assigns the authority type based on the given authority string.</span>

<span class="sd">        Args:</span>
<span class="sd">            authority (str): The authority string to be assigned.</span>

<span class="sd">        Returns:</span>
<span class="sd">            ga.AuthorityType: The corresponding authority type.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: If the authority is not supported. Only NCBI and Ensembl authorities are supported.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">authority</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">NCBI</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">NCBI</span>
        <span class="k">if</span> <span class="n">authority</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">ENSEMBL</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">ENSEMBL</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span>
            <span class="s2">&quot;Authority </span><span class="si">%s</span><span class="s2"> is not supported. Please use NCBI or Ensembl.&quot;</span><span class="p">,</span> <span class="n">authority</span>
        <span class="p">)</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Authority </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">authority</span><span class="si">}</span><span class="s2"> is not supported. Please use NCBI or Ensembl.&quot;</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Gff3.generate_genome_assembly">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.generate_genome_assembly">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_genome_assembly</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">assembly_version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">assembly_label</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">assembly_strain</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a genome assembly object.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        assembly_id (str): The ID of the assembly.</span>
<span class="sd">        assembly_version (str): The version of the assembly.</span>
<span class="sd">        assembly_label (str): The label of the assembly.</span>
<span class="sd">        assembly_strain (str, optional): The strain of the assembly. Defaults to None.</span>

<span class="sd">        Returns:</span>
<span class="sd">        ga.GenomeAssembly: The generated genome assembly object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">ga</span><span class="o">.</span><span class="n">GenomeAssembly</span><span class="p">(</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">ASSEMBLY_PREFIX</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">assembly_id</span><span class="p">,</span>
            <span class="n">in_taxon</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">id</span><span class="p">],</span>
            <span class="n">in_taxon_label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">full_name</span><span class="p">,</span>
            <span class="n">version</span><span class="o">=</span><span class="n">assembly_version</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="n">assembly_label</span><span class="p">,</span>
            <span class="n">strain</span><span class="o">=</span><span class="n">assembly_strain</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Gff3.generate_genome_annotation">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.generate_genome_annotation">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_genome_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_label</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">genome_version</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a genome annotation object.</span>

<span class="sd">        Args:</span>
<span class="sd">            genome_label (str): The label of the genome.</span>
<span class="sd">            genome_version (str): The version of the genome.</span>

<span class="sd">        Returns:</span>
<span class="sd">            ga.GenomeAnnotation: The generated genome annotation.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">ga</span><span class="o">.</span><span class="n">GenomeAnnotation</span><span class="p">(</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">BICAN_ANNOTATION_PREFIX</span> <span class="o">+</span> <span class="n">genome_label</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span>
            <span class="n">digest</span><span class="o">=</span><span class="p">[</span><span class="n">checksum</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">checksum</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">checksums</span><span class="p">],</span>
            <span class="n">content_url</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">content_url</span><span class="p">],</span>
            <span class="n">reference_assembly</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_assembly</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="n">version</span><span class="o">=</span><span class="n">genome_version</span><span class="p">,</span>
            <span class="n">in_taxon</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">id</span><span class="p">],</span>
            <span class="n">in_taxon_label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">full_name</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="n">GENOME_ANNOTATION_DESCRIPTION_FORMAT</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">authority</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">authority</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                <span class="n">taxon_scientific_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">full_name</span><span class="p">,</span>
                <span class="n">genome_version</span><span class="o">=</span><span class="n">genome_version</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">authority</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">authority</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Gff3.generate_digest">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.generate_digest">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_digest</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">hash_values</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
        <span class="n">hash_functions</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">DEFAULT_HASH</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">ga</span><span class="o">.</span><span class="n">Checksum</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates checksum digests for the GFF file using the specified hash functions.</span>

<span class="sd">        Args:</span>
<span class="sd">            hash_functions (list[str]): A list of hash functions to use for generating the digests.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list[ga.Checksum]: A list of Checksum objects containing the generated digests.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If an unsupported hash algorithm is provided.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">checksums</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">hash_type</span> <span class="ow">in</span> <span class="n">hash_functions</span><span class="p">:</span>
            <span class="c1"># Generate a UUID version 4</span>
            <span class="n">uuid_value</span> <span class="o">=</span> <span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">()</span>

            <span class="c1"># Construct a URN with the UUID</span>
            <span class="n">urn</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;urn:uuid:</span><span class="si">{</span><span class="n">uuid_value</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">hash_type</span> <span class="o">=</span> <span class="n">hash_type</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="c1"># Create a Checksum object</span>
            <span class="k">if</span> <span class="n">hash_type</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">DigestType</span><span class="o">.</span><span class="n">SHA256</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">checksums</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">ga</span><span class="o">.</span><span class="n">Checksum</span><span class="p">(</span>
                        <span class="nb">id</span><span class="o">=</span><span class="n">urn</span><span class="p">,</span>
                        <span class="n">checksum_algorithm</span><span class="o">=</span><span class="n">ga</span><span class="o">.</span><span class="n">DigestType</span><span class="o">.</span><span class="n">SHA256</span><span class="p">,</span>
                        <span class="n">value</span><span class="o">=</span><span class="n">hash_values</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;SHA256&quot;</span><span class="p">),</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">hash_type</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">DigestType</span><span class="o">.</span><span class="n">MD5</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">checksums</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">ga</span><span class="o">.</span><span class="n">Checksum</span><span class="p">(</span>
                        <span class="nb">id</span><span class="o">=</span><span class="n">urn</span><span class="p">,</span>
                        <span class="n">checksum_algorithm</span><span class="o">=</span><span class="n">ga</span><span class="o">.</span><span class="n">DigestType</span><span class="o">.</span><span class="n">MD5</span><span class="p">,</span>
                        <span class="n">value</span><span class="o">=</span><span class="n">hash_values</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;MD5&quot;</span><span class="p">),</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">hash_type</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">DigestType</span><span class="o">.</span><span class="n">SHA1</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="n">checksums</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">ga</span><span class="o">.</span><span class="n">Checksum</span><span class="p">(</span>
                        <span class="nb">id</span><span class="o">=</span><span class="n">urn</span><span class="p">,</span>
                        <span class="n">checksum_algorithm</span><span class="o">=</span><span class="n">ga</span><span class="o">.</span><span class="n">DigestType</span><span class="o">.</span><span class="n">SHA1</span><span class="p">,</span>
                        <span class="n">value</span><span class="o">=</span><span class="n">hash_values</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;SHA1&quot;</span><span class="p">),</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="s2">&quot;Hash algorithm </span><span class="si">%s</span><span class="s2"> is not supported. Please use SHA256, MD5, or SHA1.&quot;</span><span class="p">,</span>
                    <span class="n">hash_type</span><span class="p">,</span>
                <span class="p">)</span>
        <span class="k">return</span> <span class="n">checksums</span></div>


    <span class="k">def</span> <span class="nf">__get_line_count</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the line count of a file.</span>

<span class="sd">        Args:</span>
<span class="sd">            file_path (str): The path to the file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: The number of lines in the file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;wc&quot;</span><span class="p">,</span> <span class="s2">&quot;-l&quot;</span><span class="p">,</span> <span class="n">file_path</span><span class="p">],</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>  <span class="c1"># If check is True and the exit code was non-zero, it raises a CalledProcessError.</span>
        <span class="c1"># The CalledProcessError object will have the return code in the returncode attribute,</span>
        <span class="c1"># and output &amp; stderr attributes if those streams were captured.</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">line_count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>  <span class="c1"># Extract the line count from the output</span>
        <span class="k">return</span> <span class="n">line_count</span>

<div class="viewcode-block" id="Gff3.parse">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.parse">[docs]</a>
    <span class="k">def</span> <span class="nf">parse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_filter</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">DEFAULT_FEATURE_FILTER</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parses the GFF file and extracts gene annotations based on the provided feature filter.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature_filter (tuple[str]): Tuple of feature types to include in the gene annotations.</span>

<span class="sd">        Raises:</span>
<span class="sd">            FileNotFoundError: If the GFF file does not exist.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gff_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gff_file</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gff_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
            <span class="c1"># Decompress the gzip file</span>
            <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gff_file</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>
                <span class="c1"># Create a temporary file to save the decompressed data</span>
                <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
                    <span class="c1"># Copy the decompressed data to the temporary file</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">f_in</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
                    <span class="n">gff_file</span> <span class="o">=</span> <span class="n">f_out</span><span class="o">.</span><span class="n">name</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">gff_file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">gff_file</span><span class="si">}</span><span class="s2"> does not exist.&quot;</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gff_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">curr_line_num</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">progress_bar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span>
                <span class="n">total</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">__get_line_count</span><span class="p">(</span><span class="n">gff_file</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Parsing GFF3 File&quot;</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">line_raw</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
                <span class="n">line_strip</span> <span class="o">=</span> <span class="n">line_raw</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">curr_line_num</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">line_strip</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;##gff-version 3&quot;</span><span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="s1">&#39;&quot;##gff-version 3&quot; missing from the first line of the file. The given file may not be a valid GFF3 file.&#39;</span>
                    <span class="p">)</span>
                <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">line_strip</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># blank line</span>
                    <span class="k">continue</span>
                <span class="k">elif</span> <span class="n">line_strip</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;##&quot;</span><span class="p">):</span>  <span class="c1"># TODO: parse more metadata</span>
                    <span class="k">pass</span>
                <span class="k">elif</span> <span class="n">line_strip</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>  <span class="c1"># TODO: parse more metadata</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>  <span class="c1"># line may be a feature or unknown</span>
                    <span class="n">tokens</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">strip</span><span class="p">,</span> <span class="n">line_raw</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)))</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">9</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                            <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: Features are expected 9 columns, found </span><span class="si">%s</span><span class="s2">.&quot;</span><span class="p">,</span>
                            <span class="n">curr_line_num</span><span class="p">,</span>
                            <span class="nb">len</span><span class="p">(</span><span class="n">tokens</span><span class="p">),</span>
                        <span class="p">)</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">tokens</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="n">feature_filter</span>
                    <span class="p">):</span>  <span class="c1"># only look at rows that have a type that is included in feature_filter</span>
                        <span class="n">attributes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__merge_values</span><span class="p">(</span>
                            <span class="nb">tuple</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tokens</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">))</span>
                        <span class="p">)</span>
                        <span class="c1"># TODO: Write cleaner code that calls respective generate function based on the authority automatically</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_annotation</span><span class="o">.</span><span class="n">authority</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">ENSEMBL</span><span class="p">:</span>
                            <span class="n">gene_annotation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_ensembl_gene_annotation</span><span class="p">(</span>
                                <span class="n">attributes</span><span class="p">,</span> <span class="n">curr_line_num</span>
                            <span class="p">)</span>
                            <span class="k">if</span> <span class="n">gene_annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">[</span><span class="n">gene_annotation</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_annotation</span>
                        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_annotation</span><span class="o">.</span><span class="n">authority</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">AuthorityType</span><span class="o">.</span><span class="n">NCBI</span><span class="p">:</span>
                            <span class="n">gene_annotation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_ncbi_gene_annotation</span><span class="p">(</span>
                                <span class="n">attributes</span><span class="p">,</span> <span class="n">curr_line_num</span>
                            <span class="p">)</span>
                            <span class="k">if</span> <span class="n">gene_annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">[</span><span class="n">gene_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                                    <span class="n">gene_annotation</span>
                                <span class="p">)</span>
                <span class="n">progress_bar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">curr_line_num</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">progress_bar</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="Gff3.generate_ensembl_gene_annotation">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.generate_ensembl_gene_annotation">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_ensembl_gene_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attributes</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a GeneAnnotation object for Ensembl based on the provided attributes.</span>

<span class="sd">        Args:</span>
<span class="sd">            attributes (dict): A dictionary containing the attributes of the gene.</span>
<span class="sd">            curr_line_num (int): The line number of the current row in the input file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GeneAnnotation or None: The generated GeneAnnotation object if it is not a duplicate,</span>
<span class="sd">            otherwise None.</span>

<span class="sd">        Raises:</span>
<span class="sd">            None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stable_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">stable_id</span><span class="p">:</span>
            <span class="n">stable_id</span> <span class="o">=</span> <span class="n">stable_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Check and validate the name attribute</span>
        <span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>

        <span class="c1"># Check and validate the description attribute</span>
        <span class="n">description</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>

        <span class="c1"># Check and validate the biotype attribute</span>
        <span class="n">biotype</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;biotype&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>

        <span class="n">gene_annotation</span> <span class="o">=</span> <span class="n">ga</span><span class="o">.</span><span class="n">GeneAnnotation</span><span class="p">(</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">ENSEMBL_GENE_ID_PREFIX</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">stable_id</span><span class="p">,</span>
            <span class="n">source_id</span><span class="o">=</span><span class="n">stable_id</span><span class="p">,</span>
            <span class="n">symbol</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="n">description</span><span class="p">,</span>
            <span class="n">molecular_type</span><span class="o">=</span><span class="n">biotype</span><span class="p">,</span>
            <span class="n">referenced_in</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="n">in_taxon</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">id</span><span class="p">],</span>
            <span class="n">in_taxon_label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">full_name</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># handle duplicates</span>
        <span class="k">if</span> <span class="n">gene_annotation</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">gene_annotation</span>
        <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="Gff3.generate_ncbi_gene_annotation">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.generate_ncbi_gene_annotation">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_ncbi_gene_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attributes</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a GeneAnnotation object for NCBI based on the provided attributes.</span>

<span class="sd">        Args:</span>
<span class="sd">            attributes (dict): A dictionary containing the attributes of the gene.</span>
<span class="sd">            curr_line_num (int): The line number of the current row in the input file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GeneAnnotation or None: The generated GeneAnnotation object if it is not a duplicate,</span>
<span class="sd">            otherwise None.</span>

<span class="sd">        Raises:</span>
<span class="sd">            None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stable_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="s2">&quot;Dbxref&quot;</span> <span class="ow">in</span> <span class="n">attributes</span><span class="p">:</span>
            <span class="n">dbxref</span> <span class="o">=</span> <span class="p">{</span><span class="n">t</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;Dbxref&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)}</span>
            <span class="n">geneid_values</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">reference</span> <span class="ow">in</span> <span class="n">dbxref</span><span class="p">:</span>
                <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">:</span>
                    <span class="n">geneid_values</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneid_values</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">stable_id</span> <span class="o">=</span> <span class="n">geneid_values</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: No GeneAnnotation object created for this row due to missing dbxref attribute.&quot;</span><span class="p">,</span>
                <span class="n">curr_line_num</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">stable_id</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: No GeneAnnotation object created for this row due to number of GeneIDs provided in dbxref attribute is not equal to one.&quot;</span><span class="p">,</span>
                <span class="n">curr_line_num</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="c1"># Check and validate the name attribute</span>
        <span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>

        <span class="c1"># Check and validate the description attribute</span>
        <span class="n">description</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>

        <span class="c1"># Check and validate the biotype attribute</span>
        <span class="n">biotype</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__get_attribute</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="s2">&quot;gene_biotype&quot;</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">)</span>

        <span class="c1"># Parse synonyms</span>
        <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="s2">&quot;gene_synonym&quot;</span> <span class="ow">in</span> <span class="n">attributes</span><span class="p">:</span>
            <span class="n">synonyms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="p">{</span><span class="n">t</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;gene_synonym&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)}</span>
            <span class="p">)</span>
            <span class="n">synonyms</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>  <span class="c1"># note: this is not required, but it makes the output more predictable therefore easier to test</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: synonym is not set for this row&#39;s GeneAnnotation object due to missing gene_synonym attribute.&quot;</span><span class="p">,</span>
                <span class="n">curr_line_num</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="n">gene_annotation</span> <span class="o">=</span> <span class="n">ga</span><span class="o">.</span><span class="n">GeneAnnotation</span><span class="p">(</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">NCBI_GENE_ID_PREFIX</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">stable_id</span><span class="p">,</span>
            <span class="n">source_id</span><span class="o">=</span><span class="n">stable_id</span><span class="p">,</span>
            <span class="n">symbol</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="n">description</span><span class="p">,</span>
            <span class="n">molecular_type</span><span class="o">=</span><span class="n">biotype</span><span class="p">,</span>
            <span class="n">referenced_in</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="n">in_taxon</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">id</span><span class="p">],</span>
            <span class="n">in_taxon_label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">full_name</span><span class="p">,</span>
            <span class="n">synonym</span><span class="o">=</span><span class="n">synonyms</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">gene_annotation</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">[</span><span class="n">gene_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">]:</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">__resolve_ncbi_gene_annotation</span><span class="p">(</span>
                    <span class="n">gene_annotation</span><span class="p">,</span> <span class="n">curr_line_num</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">name</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">[</span><span class="n">gene_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                    <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: GeneAnnotation object with id </span><span class="si">%s</span><span class="s2"> already exists with a different name. Current name: </span><span class="si">%s</span><span class="s2">, Existing name: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">curr_line_num</span><span class="p">,</span>
                    <span class="n">stable_id</span><span class="p">,</span>
                    <span class="n">name</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">[</span><span class="n">gene_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">gene_annotation</span></div>


    <span class="k">def</span> <span class="nf">__get_attribute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attributes</span><span class="p">,</span> <span class="n">attribute_name</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the value of a specific attribute from the given attributes dictionary.</span>

<span class="sd">        Args:</span>
<span class="sd">            attributes (dict): A dictionary containing attribute names and their values.</span>
<span class="sd">            attribute_name (str): The name of the attribute to retrieve.</span>
<span class="sd">            curr_line_num (int): The current line number for logging purposes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str or None: The value of the attribute if found, None otherwise.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">value</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">attribute_name</span> <span class="ow">in</span> <span class="n">attributes</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="n">attribute_name</span><span class="p">])</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                    <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2"> not set for this row&#39;s GeneAnnotation object due to more than one </span><span class="si">%s</span><span class="s2"> provided.&quot;</span><span class="p">,</span>
                    <span class="n">curr_line_num</span><span class="p">,</span>
                    <span class="n">attribute_name</span><span class="p">,</span>
                    <span class="n">attribute_name</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">attribute_name</span> <span class="o">==</span> <span class="s2">&quot;description&quot;</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
                    <span class="sa">r</span><span class="s2">&quot;\s*\[Source.*?\]&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;&quot;</span><span class="p">,</span>
                    <span class="n">urllib</span><span class="o">.</span><span class="n">parse</span><span class="o">.</span><span class="n">unquote</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">()),</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">attributes</span><span class="p">[</span><span class="n">attribute_name</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">value</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s1">&#39;Line </span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1"> not set for this row</span><span class="se">\&#39;</span><span class="s1">s GeneAnnotation object due to value of </span><span class="si">%s</span><span class="s1"> attribute containing &quot;,&quot;.&#39;</span><span class="p">,</span>
                        <span class="n">curr_line_num</span><span class="p">,</span>
                        <span class="n">attribute_name</span><span class="p">,</span>
                        <span class="n">attribute_name</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2"> not set for this row&#39;s GeneAnnotation object due to missing </span><span class="si">%s</span><span class="s2"> attribute.&quot;</span><span class="p">,</span>
                <span class="n">curr_line_num</span><span class="p">,</span>
                <span class="n">attribute_name</span><span class="p">,</span>
                <span class="n">attribute_name</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">value</span>

    <span class="k">def</span> <span class="nf">__resolve_ncbi_gene_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_gene_annotation</span><span class="p">,</span> <span class="n">curr_line_num</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Resolves conflicts between existing and new gene annotations based on certain conditions.</span>

<span class="sd">        Args:</span>
<span class="sd">            new_gene_annotation (GeneAnnotation): The new gene annotation to be resolved.</span>
<span class="sd">            curr_line_num (int): The current line number in the file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GeneAnnotation or None: The resolved gene annotation or None if it cannot be resolved</span>
<span class="sd">                                    or None if the resolution is in favor of the existing gene</span>
<span class="sd">                                    annotation.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If duplicates cannot be resolved.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">existing_gene_annotation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="p">[</span><span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">existing_gene_annotation</span><span class="o">.</span><span class="n">description</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">description</span> <span class="ow">is</span> <span class="kc">None</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">existing_gene_annotation</span><span class="o">.</span><span class="n">description</span> <span class="ow">is</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">description</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="n">new_gene_annotation</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">existing_gene_annotation</span><span class="o">.</span><span class="n">molecular_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">molecular_type</span> <span class="ow">is</span> <span class="kc">None</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">existing_gene_annotation</span><span class="o">.</span><span class="n">molecular_type</span> <span class="ow">is</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">molecular_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="n">new_gene_annotation</span>
        <span class="k">if</span> <span class="n">existing_gene_annotation</span><span class="o">.</span><span class="n">molecular_type</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">BioType</span><span class="o">.</span><span class="n">protein_coding</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">molecular_type</span> <span class="o">==</span> <span class="n">ga</span><span class="o">.</span><span class="n">BioType</span><span class="o">.</span><span class="n">protein_coding</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">new_gene_annotation</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;Line </span><span class="si">%s</span><span class="s2">: Unable to resolve duplicates for GeneID: </span><span class="si">%s</span><span class="s2">.</span><span class="se">\n</span><span class="s2">existing gene: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">new gene: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">curr_line_num</span><span class="p">,</span>
            <span class="n">new_gene_annotation</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="n">existing_gene_annotation</span><span class="p">,</span>
            <span class="n">new_gene_annotation</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">__merge_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge values from a list of lists into a dictionary of sets.</span>

<span class="sd">        Args:</span>
<span class="sd">            t (list): A list of lists containing key-value pairs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary where each key maps to a set of values.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">lst</span> <span class="ow">in</span> <span class="n">t</span><span class="p">:</span>
            <span class="n">key</span> <span class="o">=</span> <span class="n">lst</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">lst</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">value</span><span class="p">:</span>
                <span class="n">result</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">result</span>

<div class="viewcode-block" id="Gff3.serialize_to_jsonld">
<a class="viewcode-back" href="../../../bkbit.data_translators.genome_annotation_translator.html#bkbit.data_translators.genome_annotation_translator.Gff3.serialize_to_jsonld">[docs]</a>
    <span class="k">def</span> <span class="nf">serialize_to_jsonld</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exclude_none</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">exclude_unset</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Serialize the object and either write it to the specified output file or print it to the CLI.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            exclude_none (bool): Whether to exclude None values in the output.</span>
<span class="sd">            exclude_unset (bool): Whether to exclude unset values in the output.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">organism_taxon</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                <span class="n">exclude_none</span><span class="o">=</span><span class="n">exclude_none</span><span class="p">,</span> <span class="n">exclude_unset</span><span class="o">=</span><span class="n">exclude_unset</span>
            <span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome_assembly</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                <span class="n">exclude_none</span><span class="o">=</span><span class="n">exclude_none</span><span class="p">,</span> <span class="n">exclude_unset</span><span class="o">=</span><span class="n">exclude_unset</span>
            <span class="p">),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome_annotation</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                <span class="n">exclude_none</span><span class="o">=</span><span class="n">exclude_none</span><span class="p">,</span> <span class="n">exclude_unset</span><span class="o">=</span><span class="n">exclude_unset</span>
            <span class="p">),</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">ck</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">checksums</span><span class="p">:</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ck</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span><span class="n">exclude_none</span><span class="o">=</span><span class="n">exclude_none</span><span class="p">,</span> <span class="n">exclude_unset</span><span class="o">=</span><span class="n">exclude_unset</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">ga</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ga</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span><span class="n">exclude_none</span><span class="o">=</span><span class="n">exclude_none</span><span class="p">,</span> <span class="n">exclude_unset</span><span class="o">=</span><span class="n">exclude_unset</span><span class="p">))</span>

        <span class="n">output_data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;@context&quot;</span><span class="p">:</span> <span class="s2">&quot;https://raw.githubusercontent.com/brain-bican/models/main/jsonld-context-autogen/genome_annotation.context.jsonld&quot;</span><span class="p">,</span>
            <span class="s2">&quot;@graph&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">output_data</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">))</span></div>
</div>



<span class="nd">@click</span><span class="o">.</span><span class="n">command</span><span class="p">()</span>
<span class="c1">##ARGUEMENTS##</span>
<span class="c1"># Argument #1: The URL of the GFF file</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">argument</span><span class="p">(</span><span class="s2">&quot;content_url&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>

<span class="c1">##OPTIONS##</span>
<span class="c1"># Option #1: The ID of the genome assembly</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span><span class="s2">&quot;assembly_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;-a&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
<span class="c1"># Option #2: The strain of the genome assembly</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;--assembly_strain&quot;</span><span class="p">,</span>
    <span class="s2">&quot;-s&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The strain of the genome assembly. Defaults to None.&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># Option #3: The log level</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;--log_level&quot;</span><span class="p">,</span>
    <span class="s2">&quot;-l&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The log level. Defaults to WARNING.&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># Option #4: Log to file</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;--log_to_file&quot;</span><span class="p">,</span>
    <span class="s2">&quot;-f&quot;</span><span class="p">,</span>
    <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Log to a file instead of the console.&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="k">def</span> <span class="nf">gff2jsonld</span><span class="p">(</span><span class="n">content_url</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">assembly_strain</span><span class="p">,</span> <span class="n">log_level</span><span class="p">,</span> <span class="n">log_to_file</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Creates GeneAnnotation objects from a GFF3 file and serializes them to JSON-LD format.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">gff3</span> <span class="o">=</span> <span class="n">Gff3</span><span class="p">(</span>
        <span class="n">content_url</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">assembly_strain</span><span class="p">,</span> <span class="n">log_level</span><span class="p">,</span> <span class="n">log_to_file</span>
    <span class="p">)</span>
    <span class="n">gff3</span><span class="o">.</span><span class="n">parse</span><span class="p">()</span>
    <span class="n">gff3</span><span class="o">.</span><span class="n">serialize_to_jsonld</span><span class="p">()</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">gff2jsonld</span><span class="p">()</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, BICAN.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>